oidc-client, ContextNotActiveException in new thread

### Describe the bug



It seams like Context is lost in the new thread when using Tokens from oidc-client

See also:
https://github.com/quarkusio/quarkus/discussions/22273

### Expected behavior

Usage of io.quarkus.oidc.client.Tokens should work, ie. have a value

### Actual behavior

`Usage(@Inject) of io.quarkus.oidc.client.Tokens is causing: ContextNotActiveException`

The result from running: MyResourceTest.fail:

```posh
javax.ws.rs.ProcessingException: javax.enterprise.context.ContextNotActiveException
    at org.jboss.resteasy.client.jaxrs.internal.ClientInvocation.filterRequest(ClientInvocation.java:780)
    at org.jboss.resteasy.microprofile.client.impl.MpClientInvocation.filterRequest(MpClientInvocation.java:75)
    at org.jboss.resteasy.client.jaxrs.internal.ClientInvocation.invoke(ClientInvocation.java:491)
    at org.jboss.resteasy.client.jaxrs.internal.proxy.ClientInvoker.invokeSync(ClientInvoker.java:152)
    at org.jboss.resteasy.client.jaxrs.internal.proxy.ClientInvoker.invoke(ClientInvoker.java:115)
    at org.jboss.resteasy.client.jaxrs.internal.proxy.ClientProxy.invoke(ClientProxy.java:76)
    at com.sun.proxy.$Proxy107.updatestate(Unknown Source)
    at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
    at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.base/java.lang.reflect.Method.invoke(Method.java:566)
    at org.jboss.resteasy.microprofile.client.ProxyInvocationHandler.invoke(ProxyInvocationHandler.java:146)
    at com.sun.proxy.$Proxy108.updatestate(Unknown Source)
    at org.my.StreamService.lambda$0(StreamService.java:26)
    at org.my.AbstractMeglerService.execute(AbstractMeglerService.java:17)
    at org.my.StreamService.updateState(StreamService.java:26)
    at org.my.StreamService_ClientProxy.updateState(Unknown Source)
    at org.my.QConsumer.run(QConsumer.java:35)
    at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
    at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
    at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
    at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
    at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: javax.enterprise.context.ContextNotActiveException
    at io.quarkus.arc.impl.ClientProxies.getDelegate(ClientProxies.java:46)
    at io.quarkus.oidc.client.runtime.TokensProducer_MidtNorge_ProducerMethod_produceTokens_494654c3bfa2346b017c24e2e2c4fcfc99087d0e_ClientProxy.arc$delegate(Unknown Source)
    at io.quarkus.oidc.client.runtime.TokensProducer_MidtNorge_ProducerMethod_produceTokens_494654c3bfa2346b017c24e2e2c4fcfc99087d0e_ClientProxy.getAccessToken(Unknown Source)
    at org.my.StreamAuthenticationFilterMidtNorge.filter(StreamAuthenticationFilterMidtNorge.java:21)
    at org.jboss.resteasy.client.jaxrs.internal.ClientInvocation.filterRequest(ClientInvocation.java:767)
    ... 22 more
```

### How to Reproduce?

It is based on: https://quarkus.io/guides/rest-client
I added my stuff in the package: org.my

The problem can be verified by running the test: MyResourceTest.fail

The other test: MyResourceTest.ok is doing the same, BUT in same thread (that is not via a Q, as the other test),
but it requires proper values in: application.properties to succeed.

### Output of `uname -a` or `ver`

uname -a Linux hty80028.sb1a.sparebank1.no 4.18.0-348.2.1.el8_5.x86_64 #1 SMP Mon Nov 8 13:30:15 EST 2021 x86_64 x86_64 x86_64 GNU/Linux

### Output of `java -version`

openjdk version "11.0.10" 2021-01-19 OpenJDK Runtime Environment AdoptOpenJDK (build 11.0.10+9) OpenJDK 64-Bit Server VM AdoptOpenJDK (build 11.0.10+9, mixed mode)

### GraalVM version (if different from Java)

_No response_

### Quarkus version or git rev

2.5.4.Final

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

_No response_

### Additional information

[rest-client-quickstart.zip](https://github.com/quarkusio/quarkus/files/7763420/rest-client-quickstart.zip)

@sberyozkin 
Just to be clear, the result is the same when I use: quarkus.oidc-client.early-tokens-acquisition=false

But the stacktrace is different related to the class AbstractTokensProducer which you talk about:
Here is the actual stacktrace, but I guess you already have seen this

`
javax.ws.rs.ProcessingException: java.util.concurrent.CompletionException: java.lang.InterruptedException
	at org.jboss.resteasy.client.jaxrs.internal.ClientInvocation.filterRequest(ClientInvocation.java:780)
	at org.jboss.resteasy.microprofile.client.impl.MpClientInvocation.filterRequest(MpClientInvocation.java:75)
	at org.jboss.resteasy.client.jaxrs.internal.ClientInvocation.invoke(ClientInvocation.java:491)
	at org.jboss.resteasy.client.jaxrs.internal.proxy.ClientInvoker.invokeSync(ClientInvoker.java:152)
	at org.jboss.resteasy.client.jaxrs.internal.proxy.ClientInvoker.invoke(ClientInvoker.java:115)
	at org.jboss.resteasy.client.jaxrs.internal.proxy.ClientProxy.invoke(ClientProxy.java:76)
	at com.sun.proxy.$Proxy107.updatestate(Unknown Source)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.jboss.resteasy.microprofile.client.ProxyInvocationHandler.invoke(ProxyInvocationHandler.java:146)
	at com.sun.proxy.$Proxy108.updatestate(Unknown Source)
	at org.my.StreamService.lambda$0(StreamService.java:26)
	at org.my.AbstractMeglerService.execute(AbstractMeglerService.java:17)
	at org.my.StreamService.updateState(StreamService.java:26)
	at org.my.StreamService_ClientProxy.updateState(Unknown Source)
	at org.my.QConsumer.run(QConsumer.java:33)
	at io.smallrye.context.impl.wrappers.SlowContextualRunnable.run(SlowContextualRunnable.java:19)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at io.quarkus.vertx.core.runtime.VertxCoreRecorder$13.runWith(VertxCoreRecorder.java:548)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2449)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1478)
	at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:29)
	at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:29)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: java.util.concurrent.CompletionException: java.lang.InterruptedException
	at io.smallrye.mutiny.operators.uni.UniBlockingAwait.await(UniBlockingAwait.java:73)
	at io.smallrye.mutiny.groups.UniAwait.atMost(UniAwait.java:62)
	at io.smallrye.mutiny.groups.UniAwait.indefinitely(UniAwait.java:43)
	at io.quarkus.oidc.client.runtime.AbstractTokensProducer.awaitTokens(AbstractTokensProducer.java:47)
	at io.quarkus.oidc.client.runtime.TokensProducer_MidtNorge.produceTokens(Unknown Source)
	at io.quarkus.oidc.client.runtime.TokensProducer_MidtNorge_ProducerMethod_produceTokens_494654c3bfa2346b017c24e2e2c4fcfc99087d0e_Bean.create(Unknown Source)
	at io.quarkus.oidc.client.runtime.TokensProducer_MidtNorge_ProducerMethod_produceTokens_494654c3bfa2346b017c24e2e2c4fcfc99087d0e_Bean.create(Unknown Source)
	at io.quarkus.arc.impl.RequestContext.getIfActive(RequestContext.java:68)
	at io.quarkus.arc.impl.ClientProxies.getDelegate(ClientProxies.java:27)
	at io.quarkus.oidc.client.runtime.TokensProducer_MidtNorge_ProducerMethod_produceTokens_494654c3bfa2346b017c24e2e2c4fcfc99087d0e_ClientProxy.arc$delegate(Unknown Source)
	at io.quarkus.oidc.client.runtime.TokensProducer_MidtNorge_ProducerMethod_produceTokens_494654c3bfa2346b017c24e2e2c4fcfc99087d0e_ClientProxy.getAccessToken(Unknown Source)
	at org.my.StreamAuthenticationFilterMidtNorge.filter(StreamAuthenticationFilterMidtNorge.java:21)
	at org.jboss.resteasy.client.jaxrs.internal.ClientInvocation.filterRequest(ClientInvocation.java:767)
	... 27 more
Caused by: java.lang.InterruptedException
	at java.base/java.util.concurrent.locks.AbstractQueuedSynchronizer.doAcquireSharedInterruptibly(AbstractQueuedSynchronizer.java:1040)
	at java.base/java.util.concurrent.locks.AbstractQueuedSynchronizer.acquireSharedInterruptibly(AbstractQueuedSynchronizer.java:1345)
	at java.base/java.util.concurrent.CountDownLatch.await(CountDownLatch.java:232)
	at io.smallrye.mutiny.operators.uni.UniBlockingAwait.await(UniBlockingAwait.java:61)
	... 39 more
`

@sberyozkin 
I get the same behavior if I try to use reactive, see stacktrace below.
But we are for the moment not able to use reactive for our project anyway.
`
java.util.concurrent.CompletionException: java.lang.InterruptedException
	at io.smallrye.mutiny.operators.uni.UniBlockingAwait.await(UniBlockingAwait.java:73)
	at io.smallrye.mutiny.groups.UniAwait.atMost(UniAwait.java:62)
	at io.smallrye.mutiny.groups.UniAwait.indefinitely(UniAwait.java:43)
	at io.quarkus.oidc.client.runtime.TokensHelper.initTokens(TokensHelper.java:20)
	at io.quarkus.oidc.client.runtime.AbstractTokensProducer.initTokens(AbstractTokensProducer.java:38)
	at io.quarkus.oidc.client.runtime.AbstractTokensProducer.init(AbstractTokensProducer.java:33)
	at io.quarkus.oidc.client.runtime.TokensProducer_MidtNorge_Bean.create(Unknown Source)
	at io.quarkus.oidc.client.runtime.TokensProducer_MidtNorge_Bean.create(Unknown Source)
	at io.quarkus.arc.impl.AbstractSharedContext.createInstanceHandle(AbstractSharedContext.java:96)
	at io.quarkus.arc.impl.AbstractSharedContext$1.get(AbstractSharedContext.java:29)
	at io.quarkus.arc.impl.AbstractSharedContext$1.get(AbstractSharedContext.java:26)
	at io.quarkus.arc.impl.LazyValue.get(LazyValue.java:26)
	at io.quarkus.arc.impl.ComputingCache.computeIfAbsent(ComputingCache.java:69)
	at io.quarkus.arc.impl.AbstractSharedContext.get(AbstractSharedContext.java:26)
	at io.quarkus.oidc.client.runtime.TokensProducer_MidtNorge_Bean.get(Unknown Source)
	at io.quarkus.oidc.client.runtime.TokensProducer_MidtNorge_Bean.get(Unknown Source)
	at io.quarkus.oidc.client.runtime.TokensProducer_MidtNorge_ProducerMethod_produceTokens_494654c3bfa2346b017c24e2e2c4fcfc99087d0e_Bean.create(Unknown Source)
	at io.quarkus.oidc.client.runtime.TokensProducer_MidtNorge_ProducerMethod_produceTokens_494654c3bfa2346b017c24e2e2c4fcfc99087d0e_Bean.create(Unknown Source)
	at io.quarkus.arc.impl.RequestContext.getIfActive(RequestContext.java:68)
	at io.quarkus.arc.impl.ClientProxies.getDelegate(ClientProxies.java:27)
	at io.quarkus.oidc.client.runtime.TokensProducer_MidtNorge_ProducerMethod_produceTokens_494654c3bfa2346b017c24e2e2c4fcfc99087d0e_ClientProxy.arc$delegate(Unknown Source)
	at io.quarkus.oidc.client.runtime.TokensProducer_MidtNorge_ProducerMethod_produceTokens_494654c3bfa2346b017c24e2e2c4fcfc99087d0e_ClientProxy.getAccessToken(Unknown Source)
	at org.my.StreamAuthenticationFilterMidtNorge.filter(StreamAuthenticationFilterMidtNorge.java:21)
	at org.jboss.resteasy.reactive.client.handlers.ClientRequestFilterRestHandler.handle(ClientRequestFilterRestHandler.java:24)
	at org.jboss.resteasy.reactive.client.handlers.ClientRequestFilterRestHandler.handle(ClientRequestFilterRestHandler.java:9)
	at org.jboss.resteasy.reactive.common.core.AbstractResteasyReactiveContext.run(AbstractResteasyReactiveContext.java:141)
	at org.jboss.resteasy.reactive.client.impl.AsyncInvokerImpl.performRequestInternal(AsyncInvokerImpl.java:262)
	at org.jboss.resteasy.reactive.client.impl.AsyncInvokerImpl.performRequestInternal(AsyncInvokerImpl.java:252)
	at org.jboss.resteasy.reactive.client.impl.AsyncInvokerImpl.method(AsyncInvokerImpl.java:212)
	at org.jboss.resteasy.reactive.client.impl.InvocationBuilderImpl.method(InvocationBuilderImpl.java:324)
	at org.my.StreamClientMidtNorge$$QuarkusRestClientInterface.updatestate(Unknown Source)
	at org.my.StreamService.lambda$0(StreamService.java:26)
	at org.my.AbstractMeglerService.execute(AbstractMeglerService.java:17)
	at org.my.StreamService.updateState(StreamService.java:26)
	at org.my.StreamService_ClientProxy.updateState(Unknown Source)
	at org.my.QConsumer.run(QConsumer.java:33)
	at io.smallrye.context.impl.wrappers.SlowContextualRunnable.run(SlowContextualRunnable.java:19)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at io.quarkus.vertx.core.runtime.VertxCoreRecorder$13.runWith(VertxCoreRecorder.java:548)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2449)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1478)
	at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:29)
	at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:29)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: java.lang.InterruptedException
	at java.base/java.util.concurrent.locks.AbstractQueuedSynchronizer.doAcquireSharedInterruptibly(AbstractQueuedSynchronizer.java:1040)
	at java.base/java.util.concurrent.locks.AbstractQueuedSynchronizer.acquireSharedInterruptibly(AbstractQueuedSynchronizer.java:1345)
	at java.base/java.util.concurrent.CountDownLatch.await(CountDownLatch.java:232)
	at io.smallrye.mutiny.operators.uni.UniBlockingAwait.await(UniBlockingAwait.java:61)
	... 45 more
`

@sberyozkin 
I get the same behavior if I try without a Q, but in another thread
See stacktrace below

Required change:
`
MyResource
@Path("/ok")
    @GET
    public void ok() {
        //service.updateState();
        CompletableFuture.runAsync(() -> service.updateState(), managedExecutor); // managedExecutor injected
    }


java.util.concurrent.CompletionException: java.lang.InterruptedException
	at io.smallrye.mutiny.operators.uni.UniBlockingAwait.await(UniBlockingAwait.java:73)
	at io.smallrye.mutiny.groups.UniAwait.atMost(UniAwait.java:62)
	at io.smallrye.mutiny.groups.UniAwait.indefinitely(UniAwait.java:43)
	at io.quarkus.oidc.client.runtime.TokensHelper.initTokens(TokensHelper.java:20)
	at io.quarkus.oidc.client.runtime.AbstractTokensProducer.initTokens(AbstractTokensProducer.java:38)
	at io.quarkus.oidc.client.runtime.AbstractTokensProducer.init(AbstractTokensProducer.java:33)
	at io.quarkus.oidc.client.runtime.TokensProducer_MidtNorge_Bean.create(Unknown Source)
	at io.quarkus.oidc.client.runtime.TokensProducer_MidtNorge_Bean.create(Unknown Source)
	at io.quarkus.arc.impl.AbstractSharedContext.createInstanceHandle(AbstractSharedContext.java:96)
	at io.quarkus.arc.impl.AbstractSharedContext$1.get(AbstractSharedContext.java:29)
	at io.quarkus.arc.impl.AbstractSharedContext$1.get(AbstractSharedContext.java:26)
	at io.quarkus.arc.impl.LazyValue.get(LazyValue.java:26)
	at io.quarkus.arc.impl.ComputingCache.computeIfAbsent(ComputingCache.java:69)
	at io.quarkus.arc.impl.AbstractSharedContext.get(AbstractSharedContext.java:26)
	at io.quarkus.oidc.client.runtime.TokensProducer_MidtNorge_Bean.get(Unknown Source)
	at io.quarkus.oidc.client.runtime.TokensProducer_MidtNorge_Bean.get(Unknown Source)
	at io.quarkus.oidc.client.runtime.TokensProducer_MidtNorge_ProducerMethod_produceTokens_494654c3bfa2346b017c24e2e2c4fcfc99087d0e_Bean.create(Unknown Source)
	at io.quarkus.oidc.client.runtime.TokensProducer_MidtNorge_ProducerMethod_produceTokens_494654c3bfa2346b017c24e2e2c4fcfc99087d0e_Bean.create(Unknown Source)
	at io.quarkus.arc.impl.RequestContext.getIfActive(RequestContext.java:68)
	at io.quarkus.arc.impl.ClientProxies.getDelegate(ClientProxies.java:27)
	at io.quarkus.oidc.client.runtime.TokensProducer_MidtNorge_ProducerMethod_produceTokens_494654c3bfa2346b017c24e2e2c4fcfc99087d0e_ClientProxy.arc$delegate(Unknown Source)
	at io.quarkus.oidc.client.runtime.TokensProducer_MidtNorge_ProducerMethod_produceTokens_494654c3bfa2346b017c24e2e2c4fcfc99087d0e_ClientProxy.getAccessToken(Unknown Source)
	at org.my.StreamAuthenticationFilterMidtNorge.filter(StreamAuthenticationFilterMidtNorge.java:21)
	at org.jboss.resteasy.reactive.client.handlers.ClientRequestFilterRestHandler.handle(ClientRequestFilterRestHandler.java:24)
	at org.jboss.resteasy.reactive.client.handlers.ClientRequestFilterRestHandler.handle(ClientRequestFilterRestHandler.java:9)
	at org.jboss.resteasy.reactive.common.core.AbstractResteasyReactiveContext.run(AbstractResteasyReactiveContext.java:141)
	at org.jboss.resteasy.reactive.client.impl.AsyncInvokerImpl.performRequestInternal(AsyncInvokerImpl.java:262)
	at org.jboss.resteasy.reactive.client.impl.AsyncInvokerImpl.performRequestInternal(AsyncInvokerImpl.java:252)
	at org.jboss.resteasy.reactive.client.impl.AsyncInvokerImpl.method(AsyncInvokerImpl.java:212)
	at org.jboss.resteasy.reactive.client.impl.InvocationBuilderImpl.method(InvocationBuilderImpl.java:324)
	at org.my.StreamClientMidtNorge$$QuarkusRestClientInterface.updatestate(Unknown Source)
	at org.my.StreamService.lambda$0(StreamService.java:26)
	at org.my.AbstractMeglerService.execute(AbstractMeglerService.java:17)
	at org.my.StreamService.updateState(StreamService.java:26)
	at org.my.StreamService_ClientProxy.updateState(Unknown Source)
	at org.my.MyResource.lambda$0(MyResource.java:30)
	at java.base/java.util.concurrent.CompletableFuture$AsyncRun.run(CompletableFuture.java:1736)
	at io.smallrye.context.impl.wrappers.SlowContextualRunnable.run(SlowContextualRunnable.java:19)
	at io.quarkus.vertx.core.runtime.VertxCoreRecorder$13.runWith(VertxCoreRecorder.java:543)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2449)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1452)
	at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:29)
	at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:29)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: java.lang.InterruptedException
	at java.base/java.util.concurrent.locks.AbstractQueuedSynchronizer.doAcquireSharedInterruptibly(AbstractQueuedSynchronizer.java:1040)
	at java.base/java.util.concurrent.locks.AbstractQueuedSynchronizer.acquireSharedInterruptibly(AbstractQueuedSynchronizer.java:1345)
	at java.base/java.util.concurrent.CountDownLatch.await(CountDownLatch.java:232)
	at io.smallrye.mutiny.operators.uni.UniBlockingAwait.await(UniBlockingAwait.java:61)
	... 44 more
`

--- a/integration-tests/oidc-client/src/main/java/io/quarkus/it/keycloak/FrontendResource.java
+++ b/integration-tests/oidc-client/src/main/java/io/quarkus/it/keycloak/FrontendResource.java
@@ -4,6 +4,7 @@ import javax.inject.Inject;
 import javax.ws.rs.GET;
 import javax.ws.rs.Path;
 
+import org.eclipse.microprofile.context.ManagedExecutor;
 import org.eclipse.microprofile.rest.client.inject.RestClient;
 
 @Path("/frontend")
@@ -14,16 +15,50 @@ public class FrontendResource {
 
     @Inject
     @RestClient
+    ProtectedResourceServiceNamedOidcClient protectedResourceServiceNamedOidcClient;
+
+    @Inject
+    @RestClient
     ProtectedResourceServiceOidcClient protectedResourceServiceRegisterProvider;
 
     @Inject
     @RestClient
     ProtectedResourceServiceNoOidcClient protectedResourceServiceNoOidcClient;
 
+    @Inject
+    ManagedExecutor managedExecutor;
+
+    private Object lock = new Object();
+    private volatile String userName;
+
     @GET
     @Path("userOidcClient")
     public String userNameOidcClient() {
         return protectedResourceServiceOidcClient.getUserName();
+    }
+
+    @GET
+    @Path("userOidcClientManagedExecutor")
+    public String userNameOidcClientManagedExecutor() throws Exception {
+        managedExecutor.execute(new Runnable() {
+
+            @Override
+            public void run() {
+                String userNameResponse = protectedResourceServiceNamedOidcClient.getUserName();
+                synchronized (lock) {
+                    userName = userNameResponse;
+                    lock.notify();
+                }
+            }
+
+        });
+
+        synchronized (lock) {
+            while (userName == null) {
+                lock.wait();
+            }
+        }
+        return userName;
     }
 
     @GET

--- a/integration-tests/oidc-client/src/main/java/io/quarkus/it/keycloak/OidcClientRequestCustomFilter.java
+++ b/integration-tests/oidc-client/src/main/java/io/quarkus/it/keycloak/OidcClientRequestCustomFilter.java
@@ -1,0 +1,26 @@
+package io.quarkus.it.keycloak;
+
+import java.io.IOException;
+
+import javax.annotation.Priority;
+import javax.inject.Inject;
+import javax.ws.rs.Priorities;
+import javax.ws.rs.client.ClientRequestContext;
+import javax.ws.rs.client.ClientRequestFilter;
+import javax.ws.rs.core.HttpHeaders;
+
+import io.quarkus.oidc.client.NamedOidcClient;
+import io.quarkus.oidc.client.Tokens;
+
+@Priority(Priorities.AUTHENTICATION)
+public class OidcClientRequestCustomFilter implements ClientRequestFilter {
+
+    @Inject
+    @NamedOidcClient("named")
+    Tokens grantTokens;
+
+    @Override
+    public void filter(ClientRequestContext requestContext) throws IOException {
+        requestContext.getHeaders().add(HttpHeaders.AUTHORIZATION, "Bearer " + grantTokens.getAccessToken());
+    }
+}

--- a/integration-tests/oidc-client/src/main/java/io/quarkus/it/keycloak/ProtectedResourceServiceNamedOidcClient.java
+++ b/integration-tests/oidc-client/src/main/java/io/quarkus/it/keycloak/ProtectedResourceServiceNamedOidcClient.java
@@ -1,0 +1,16 @@
+package io.quarkus.it.keycloak;
+
+import javax.ws.rs.GET;
+import javax.ws.rs.Path;
+
+import org.eclipse.microprofile.rest.client.annotation.RegisterProvider;
+import org.eclipse.microprofile.rest.client.inject.RegisterRestClient;
+
+@RegisterRestClient
+@RegisterProvider(OidcClientRequestCustomFilter.class)
+@Path("/")
+public interface ProtectedResourceServiceNamedOidcClient {
+
+    @GET
+    String getUserName();
+}

--- a/integration-tests/oidc-client/src/main/resources/application.properties
+++ b/integration-tests/oidc-client/src/main/resources/application.properties
@@ -9,8 +9,16 @@ quarkus.oidc-client.grant.type=password
 quarkus.oidc-client.grant-options.password.username=alice
 quarkus.oidc-client.grant-options.password.password=alice
 
+quarkus.oidc-client.named.auth-server-url=${quarkus.oidc.auth-server-url}
+quarkus.oidc-client.named.client-id=${quarkus.oidc.client-id}
+quarkus.oidc-client.named.credentials.secret=${quarkus.oidc.credentials.secret}
+quarkus.oidc-client.named.grant.type=password
+quarkus.oidc-client.named.grant-options.password.username=alice
+quarkus.oidc-client.named.grant-options.password.password=alice
+
 io.quarkus.it.keycloak.ProtectedResourceServiceRegisterProvider/mp-rest/url=http://localhost:8081/protected
 io.quarkus.it.keycloak.ProtectedResourceServiceOidcClient/mp-rest/url=http://localhost:8081/protected
+io.quarkus.it.keycloak.ProtectedResourceServiceNamedOidcClient/mp-rest/url=http://localhost:8081/protected
 io.quarkus.it.keycloak.ProtectedResourceServiceNoOidcClient/mp-rest/url=http://localhost:8081/protected
 
 quarkus.tls.trust-all=true

--- a/integration-tests/oidc-client/src/test/java/io/quarkus/it/keycloak/OidcClientTest.java
+++ b/integration-tests/oidc-client/src/test/java/io/quarkus/it/keycloak/OidcClientTest.java
@@ -37,6 +37,14 @@ public class OidcClientTest {
     }
 
     @Test
+    public void testGetUserNameOidcClientManagedExecutor() {
+        RestAssured.when().get("/frontend/userOidcClientManagedExecutor")
+                .then()
+                .statusCode(200)
+                .body(equalTo("alice"));
+    }
+
+    @Test
     public void testGetUserOidcClientNameAndRefreshTokens() {
         RestAssured.when().get("/frontend/userOidcClient")
                 .then()

